core/src/main/java/org/lflang/federated/generator/FederateInstance.java

```diff
@@ -702,49 +697,103 @@ public String toString() {
   }
 
   /**
-   * Find the nearest (shortest) path to a physical action trigger from this 'reaction' in terms of
-   * minimum delay.
+   * Return the shortest total delay from an upstream physical action to the specified output port
+   * or TimeValue.MAX_VALUE if there is no upstream physical action.
    *
-   * @param reaction The reaction to start with
-   * @return The minimum delay found to the nearest physical action and TimeValue.MAX_VALUE
-   *     otherwise
-   */
-  public TimeValue findNearestPhysicalActionTrigger(ReactionInstance reaction) {
-    TimeValue minDelay = TimeValue.MAX_VALUE;
-    for (TriggerInstance<? extends Variable> trigger : reaction.triggers) {
-      if (trigger.getDefinition() instanceof Action action) {
-        ActionInstance actionInstance = (ActionInstance) trigger;
-        if (action.getOrigin() == ActionOrigin.PHYSICAL) {
-          if (actionInstance.getMinDelay().isEarlierThan(minDelay)) {
-            minDelay = actionInstance.getMinDelay();
-          }
-        } else if (action.getOrigin() == ActionOrigin.LOGICAL) {
-          // Logical action
-          // Follow it upstream inside the reactor
-          for (ReactionInstance uReaction : actionInstance.getDependsOnReactions()) {
-            // Avoid a loop
-            if (!Objects.equal(uReaction, reaction)) {
-              TimeValue uMinDelay =
-                  actionInstance.getMinDelay().add(findNearestPhysicalActionTrigger(uReaction));
-              if (uMinDelay.isEarlierThan(minDelay)) {
-                minDelay = uMinDelay;
-              }
-            }
-          }
-        }
+   * @param output The output port.
+   */
+  private TimeValue minDelayFromPhysicalActionTo(
+      Map<ReactionInstance, TimeValue> visited, PortInstance output) {
```

Comment:
```suggestion
   * @param port A port (either input or output) wrt which the min delay is a calculated.
   */
  private TimeValue minDelayFromPhysicalActionTo(
      Map<ReactionInstance, TimeValue> visited, PortInstance port) {
```

I suggest this name change because this function can be called on input ports a few lines down in `var uDelay = minDelayFromPhysicalActionTo(visited, upstreamPort.instance);`